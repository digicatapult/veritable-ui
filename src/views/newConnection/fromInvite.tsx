import Html from '@kitajs/html'
import { singleton } from 'tsyringe'
import { CompanyProfile } from '../../models/companyHouseEntity.js'
import { BASE_64_URL, pinCodeRegex } from '../../models/strings.js'
import { Page } from '../common.js'
import { FormFeedback, NewConnectionTemplates } from './base.js'

export type FromInviteFormStage = 'invite' | 'pin' | 'success'
export type FormInviteProps = {
  feedback: FormFeedback
  formStage: FromInviteFormStage
  pin?: string
  invite?: BASE_64_URL
  company?: CompanyProfile
}

@singleton()
export class FromInviteTemplates extends NewConnectionTemplates {
  constructor() {
    super()
  }

  public fromInviteFormPage = (feedback: FormFeedback, pin?: string) => {
    return (
      <Page
        title="Veritable - New Connection"
        heading="New Connection"
        headerLinks={[
          { name: 'Connections', url: '/connection' },
          { name: 'Add from Invitation', url: '/connection/new?fromInvite=true' },
        ]}
        stylesheets={['new-invite.css']}
      >
        <div class="connections header">
          <span>Invite New Connection</span>
        </div>
        <div class="card-body ">
          <this.fromInviteForm pin={pin} feedback={feedback} formStage="invite" />
        </div>
      </Page>
    )
  }

  public fromInviteForm = (props: FormInviteProps): JSX.Element => {
    switch (props.formStage) {
      case 'invite':
        return <this.fromInviteInvite {...props}></this.fromInviteInvite>
      case 'pin':
        return <this.newInvitePin {...props}></this.newInvitePin>
      case 'success':
        return <this.fromInviteSuccess {...props}></this.fromInviteSuccess>
    }
  }

  private fromInviteInvite = (props: FormInviteProps): JSX.Element => {
    return (
      <this.newConnectionForm
        submitRoute="receive-invitation"
        feedback={props.feedback}
        progressStep={1}
        progressStepCount={3}
        actions={[
          { type: 'link', text: 'Cancel', href: '/connection' },
          { type: 'submit', value: 'createConnection', text: 'Submit' },
        ]}
      >
        <div id="from-invite-invite-input" class="accented-container">
          <textarea
            name="invite"
            placeholder="Invitation Text"
            required
            hx-get="/connection/new/verify-invite"
            hx-trigger="keyup changed delay:200ms, change, load"
            hx-target="#new-connection-feedback"
            hx-select="#new-connection-feedback"
            hx-swap="outerHTML"
          >
            {props.invite ? Html.escapeHtml(props.invite) : ''}
          </textarea>
        </div>
      </this.newConnectionForm>
    )
  }

  public newInvitePin = (props: FormInviteProps): JSX.Element => {
    return (
      <this.newConnectionForm
        submitRoute="pin-submission"
        feedback={props.feedback}
        progressStep={2}
        progressStepCount={3}
        actions={[
          { type: 'link', text: 'Fill In Later', href: '/connection' },
          { type: 'submit', value: 'pin-submit', text: 'Continue' },
        ]}
      >
        <div class="accented-container">
          <p>Please enter the verification code from the physical letter</p>
          <input
            name="invite"
            value={(props.feedback.type === 'success' && props.feedback.invite) || ''}
            type="hidden"
          />
          <input
            id="from-invite-invite-input-pin"
            name="pin"
            class="new-connection-input-field"
            placeholder="Code"
            required
            hx-trigger="keyup changed delay:500ms"
            value={props.pin || ''}
            type="number"
            hx-target="#new-connection-feedback"
            hx-select="#new-connection-feedback"
            hx-swap="outerHTML"
            pattern={pinCodeRegex.source}
            minlength={6}
            maxlength={6}
          />
        </div>
      </this.newConnectionForm>
    )
  }

  private fromInviteSuccess = (props: FormInviteProps): JSX.Element => {
    return (
      <this.newConnectionForm
        submitRoute="receive-invitation"
        feedback={props.feedback}
        progressStep={3}
        progressStepCount={3}
        actions={[{ type: 'link', text: 'Back To Home', href: '/connection' }]}
      >
        <div id="new-invite-confirmation-text">
          <p>
            Your connection has been established, but still needs to be verified. You should receive a verification
            letter at your registered business with instructions on how to do this. A reciprocal verification request
            has been sent in the post on your behalf to the address on the right to verify their identity
          </p>
        </div>
      </this.newConnectionForm>
    )
  }
}
