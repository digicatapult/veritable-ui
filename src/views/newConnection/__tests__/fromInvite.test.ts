import { expect } from 'chai'
import { describe, it } from 'mocha'

import { FromInviteTemplates } from '../fromInvite.js'
import { testErrorTargetBox, testMessageTargetBox, testSuccessTargetBox } from './fixtures.js'

describe('FromInviteTemplates', () => {
  describe('show form', () => {
    it('should render form with a error message and invalid response', async () => {
      const templates = new FromInviteTemplates()
      const rendered = await templates.fromInviteForm({ feedback: testMessageTargetBox })
      expect(rendered).to.matchSnapshot()
    })

    it('should render a web page with the a form in an empty state', async () => {
      const templates = new FromInviteTemplates()
      const rendered = await templates.fromInviteFormPage(testSuccessTargetBox)
      expect(rendered).to.matchSnapshot()
    })

    it('should render a web page with the a form in an empty state', async () => {
      const templates = new FromInviteTemplates()
      const rendered = await templates.fromInviteFormPage(testMessageTargetBox)
      expect(rendered).to.matchSnapshot()
    })

    it('should render a web page with the a form in an error state', async () => {
      const templates = new FromInviteTemplates()
      const rendered = await templates.fromInviteFormPage(testErrorTargetBox)
      expect(rendered).to.matchSnapshot()
    })
  })
})
