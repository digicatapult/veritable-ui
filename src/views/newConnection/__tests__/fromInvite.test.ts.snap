// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromInviteTemplates show form should render a web page with the a form in an empty state 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><script src=\\"/lib/htmx.org/htmx.min.js\\"></script><script src=\\"/lib/htmx-ext-json-enc/json-enc.js\\"></script><script src=\\"/public/scripts/auth-redirect.js\\"></script><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon-dark.ico\\" media=\\"(prefers-color-scheme: dark)\\"/><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon.ico\\" media=\\"(prefers-color-scheme: light), (prefers-color-scheme: no-preference)\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/main.css\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/connection.css\\"/><title>Veritable - New Connection</title></head><body hx-ext=\\"json-enc\\"><nav id=\\"side-bar\\"><img src=\\"/public/images/logo-square.svg\\"/><div class=\\"profile-icon\\" style=\\"background-color:#38b6ff;\\">A</div><a title=\\"categories\\" href=\\"/\\" style=\\"--background-image:url('/public/images/category.svg');\\"></a><a title=\\"queries\\" href=\\"/queries\\" style=\\"--background-image:url('/public/images/query.svg');\\"></a><a title=\\"credentials\\" href=\\"/credentials\\" style=\\"--background-image:url('/public/images/credential.svg');\\"></a><a title=\\"connections\\" href=\\"/connection\\" data-active style=\\"--background-image:url('/public/images/connection.svg');\\"></a><a title=\\"settings\\" href=\\"/settings\\" style=\\"--background-image:url('/public/images/setting.svg');\\"></a></nav><main><div id=\\"content-header\\"><h1>New Connection</h1><div id=\\"content-header-nav\\"><a title=\\"home\\" class=\\"nav icon\\" href=\\"/\\"></a><a title=\\"Connections\\" href=\\"/connection\\">Connections</a><a title=\\"Add from Invitation\\" href=\\"/connection/new?fromInvite=true\\">Add from Invitation</a></div></div><div id=\\"content-main\\"><div class=\\"connections header\\"><span>Invite New Connection</span></div><form id=\\"new-connection-form\\" hx-post=\\"/connection/new/receive-invitation\\" hx-swap=\\"outerHTML\\" hx-select=\\"#new-connection-form\\"><div id=\\"new-connection-progress\\" style=\\"--progress-percent: 0.3333333333333333\\"><div id=\\"new-connection-progress-text\\"><span>Step 1</span><span> of 3</span></div><div id=\\"new-connection-progress-bar\\"></div></div><div id=\\"new-connection-invite-input\\" class=\\"accented-container\\"><textarea name=\\"invite\\" placeholder=\\"Invitation Text\\" required hx-get=\\"/connection/new/verify-invite\\" hx-trigger=\\"keyup changed delay:20ms, change, load\\" hx-target=\\"#new-connection-feedback\\" hx-select=\\"#new-connection-feedback\\" hx-swap=\\"outerHTML\\"></textarea></div><div id=\\"new-connection-feedback\\" class=\\"accented-container feedback-positive\\"><div><p><span>Registered Office Address</span><span class=\\"address-line\\">DIGITAL CATAPULT,&nbsp</span><span class=\\"address-line\\">Level 9, 101 Euston Road,&nbsp</span><span class=\\"address-line\\">London,&nbsp</span><span class=\\"address-line\\">NW1 2RA</span></p><p><span>Company Status</span>active</p></div><div><img src=\\"/public/images/check.svg\\" alt=\\"Description of the image\\"/></div></div><div id=\\"new-connection-actions\\"><a class=\\"button \\" href=\\"/connection\\" data-variant=\\"outlined\\"><span>Cancel</span></a><button class=\\"button  \\" data-variant=\\"filled\\" type=\\"submit\\" name=\\"action\\" value=\\"createConnection\\"><span>Submit</span></button></div></form></div></main></body></html>"`;

exports[`FromInviteTemplates show form should render a web page with the a form in an empty state 2`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><script src=\\"/lib/htmx.org/htmx.min.js\\"></script><script src=\\"/lib/htmx-ext-json-enc/json-enc.js\\"></script><script src=\\"/public/scripts/auth-redirect.js\\"></script><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon-dark.ico\\" media=\\"(prefers-color-scheme: dark)\\"/><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon.ico\\" media=\\"(prefers-color-scheme: light), (prefers-color-scheme: no-preference)\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/main.css\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/connection.css\\"/><title>Veritable - New Connection</title></head><body hx-ext=\\"json-enc\\"><nav id=\\"side-bar\\"><img src=\\"/public/images/logo-square.svg\\"/><div class=\\"profile-icon\\" style=\\"background-color:#38b6ff;\\">A</div><a title=\\"categories\\" href=\\"/\\" style=\\"--background-image:url('/public/images/category.svg');\\"></a><a title=\\"queries\\" href=\\"/queries\\" style=\\"--background-image:url('/public/images/query.svg');\\"></a><a title=\\"credentials\\" href=\\"/credentials\\" style=\\"--background-image:url('/public/images/credential.svg');\\"></a><a title=\\"connections\\" href=\\"/connection\\" data-active style=\\"--background-image:url('/public/images/connection.svg');\\"></a><a title=\\"settings\\" href=\\"/settings\\" style=\\"--background-image:url('/public/images/setting.svg');\\"></a></nav><main><div id=\\"content-header\\"><h1>New Connection</h1><div id=\\"content-header-nav\\"><a title=\\"home\\" class=\\"nav icon\\" href=\\"/\\"></a><a title=\\"Connections\\" href=\\"/connection\\">Connections</a><a title=\\"Add from Invitation\\" href=\\"/connection/new?fromInvite=true\\">Add from Invitation</a></div></div><div id=\\"content-main\\"><div class=\\"connections header\\"><span>Invite New Connection</span></div><form id=\\"new-connection-form\\" hx-post=\\"/connection/new/receive-invitation\\" hx-swap=\\"outerHTML\\" hx-select=\\"#new-connection-form\\"><div id=\\"new-connection-progress\\" style=\\"--progress-percent: 0.3333333333333333\\"><div id=\\"new-connection-progress-text\\"><span>Step 1</span><span> of 3</span></div><div id=\\"new-connection-progress-bar\\"></div></div><div id=\\"new-connection-invite-input\\" class=\\"accented-container\\"><textarea name=\\"invite\\" placeholder=\\"Invitation Text\\" required hx-get=\\"/connection/new/verify-invite\\" hx-trigger=\\"keyup changed delay:20ms, change, load\\" hx-target=\\"#new-connection-feedback\\" hx-select=\\"#new-connection-feedback\\" hx-swap=\\"outerHTML\\"></textarea></div><div id=\\"new-connection-feedback\\" class=\\"accented-container feedback-neutral\\">This is a message</div><div id=\\"new-connection-actions\\"><a class=\\"button \\" href=\\"/connection\\" data-variant=\\"outlined\\"><span>Cancel</span></a><button class=\\"button  \\" data-variant=\\"filled\\" type=\\"submit\\" name=\\"action\\" value=\\"createConnection\\"><span>Submit</span></button></div></form></div></main></body></html>"`;

exports[`FromInviteTemplates show form should render a web page with the a form in an error state 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><script src=\\"/lib/htmx.org/htmx.min.js\\"></script><script src=\\"/lib/htmx-ext-json-enc/json-enc.js\\"></script><script src=\\"/public/scripts/auth-redirect.js\\"></script><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon-dark.ico\\" media=\\"(prefers-color-scheme: dark)\\"/><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon.ico\\" media=\\"(prefers-color-scheme: light), (prefers-color-scheme: no-preference)\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/main.css\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/connection.css\\"/><title>Veritable - New Connection</title></head><body hx-ext=\\"json-enc\\"><nav id=\\"side-bar\\"><img src=\\"/public/images/logo-square.svg\\"/><div class=\\"profile-icon\\" style=\\"background-color:#38b6ff;\\">A</div><a title=\\"categories\\" href=\\"/\\" style=\\"--background-image:url('/public/images/category.svg');\\"></a><a title=\\"queries\\" href=\\"/queries\\" style=\\"--background-image:url('/public/images/query.svg');\\"></a><a title=\\"credentials\\" href=\\"/credentials\\" style=\\"--background-image:url('/public/images/credential.svg');\\"></a><a title=\\"connections\\" href=\\"/connection\\" data-active style=\\"--background-image:url('/public/images/connection.svg');\\"></a><a title=\\"settings\\" href=\\"/settings\\" style=\\"--background-image:url('/public/images/setting.svg');\\"></a></nav><main><div id=\\"content-header\\"><h1>New Connection</h1><div id=\\"content-header-nav\\"><a title=\\"home\\" class=\\"nav icon\\" href=\\"/\\"></a><a title=\\"Connections\\" href=\\"/connection\\">Connections</a><a title=\\"Add from Invitation\\" href=\\"/connection/new?fromInvite=true\\">Add from Invitation</a></div></div><div id=\\"content-main\\"><div class=\\"connections header\\"><span>Invite New Connection</span></div><form id=\\"new-connection-form\\" hx-post=\\"/connection/new/receive-invitation\\" hx-swap=\\"outerHTML\\" hx-select=\\"#new-connection-form\\"><div id=\\"new-connection-progress\\" style=\\"--progress-percent: 0.3333333333333333\\"><div id=\\"new-connection-progress-text\\"><span>Step 1</span><span> of 3</span></div><div id=\\"new-connection-progress-bar\\"></div></div><div id=\\"new-connection-invite-input\\" class=\\"accented-container\\"><textarea name=\\"invite\\" placeholder=\\"Invitation Text\\" required hx-get=\\"/connection/new/verify-invite\\" hx-trigger=\\"keyup changed delay:20ms, change, load\\" hx-target=\\"#new-connection-feedback\\" hx-select=\\"#new-connection-feedback\\" hx-swap=\\"outerHTML\\"></textarea></div><div id=\\"new-connection-feedback\\" class=\\"accented-container feedback-negative\\">This is a test error message</div><div id=\\"new-connection-actions\\"><a class=\\"button \\" href=\\"/connection\\" data-variant=\\"outlined\\"><span>Cancel</span></a><button class=\\"button  \\" data-variant=\\"filled\\" type=\\"submit\\" name=\\"action\\" value=\\"createConnection\\"><span>Submit</span></button></div></form></div></main></body></html>"`;

exports[`FromInviteTemplates show form should render form with a error message and invalid response 1`] = `"<form id=\\"new-connection-form\\" hx-post=\\"/connection/new/receive-invitation\\" hx-swap=\\"outerHTML\\" hx-select=\\"#new-connection-form\\"><div id=\\"new-connection-progress\\" style=\\"--progress-percent: 0.3333333333333333\\"><div id=\\"new-connection-progress-text\\"><span>Step 1</span><span> of 3</span></div><div id=\\"new-connection-progress-bar\\"></div></div><div id=\\"new-connection-invite-input\\" class=\\"accented-container\\"><textarea name=\\"invite\\" placeholder=\\"Invitation Text\\" required hx-get=\\"/connection/new/verify-invite\\" hx-trigger=\\"keyup changed delay:20ms, change, load\\" hx-target=\\"#new-connection-feedback\\" hx-select=\\"#new-connection-feedback\\" hx-swap=\\"outerHTML\\"></textarea></div><div id=\\"new-connection-feedback\\" class=\\"accented-container feedback-neutral\\">This is a message</div><div id=\\"new-connection-actions\\"><a class=\\"button \\" href=\\"/connection\\" data-variant=\\"outlined\\"><span>Cancel</span></a><button class=\\"button  \\" data-variant=\\"filled\\" type=\\"submit\\" name=\\"action\\" value=\\"createConnection\\" text=\\"Submit\\" style=\\"filled\\"><span>Submit</span></button></div></form>"`;
